### PR TITLE
Support source code links in publication lists when the ‘code’ field is present in the corresponding BibTeX entries

### PR DESCRIPTION
**Pull request by dcallega**
Pick up URLs from the `code` fields of BibTeX entries, and render
them as links in the row of extras next to each publication entry.

The custom BibTeX `code` field is to be used solely for providing
an (external) link to the source code of the experiments published
in the paper (e.g., a link to a GitHub repository page).

### CODE REVIEW

File: layouts/partials/bib/entry/extras.html [hunk 1]
```diff
@@ -14,6 +14,9 @@
 {{- with .entry.doi -}}
   <span class="{{ $class }} pubref"><a href="https://doi.org/{{ . }}" target="_blank" rel="noopener">Release</a></span>
 {{- end -}}
+{{- with .entry.code -}}
+  <span class="{{ $class }} pubref"><a href="{{ . }}" target="_blank" rel="noopener">Code</a></span>
```
Comment (igorburago) [inline layouts/partials/bib/entry/extras.html]: Change the `pubref` class to, say, `coderef`:

```suggestion
  <span class="{{ $class }} coderef"><a href="{{ . }}" target="_blank" rel="noopener">Code</a></span>
```

The classes of different extras’ `<span>`s (beyond the common
`$class`) are supposed to be different to allow for their potential
differentiation in styling.

File: layouts/partials/bib/entry/extras.html [hunk 2]
```diff
@@ -14,6 +14,9 @@
...
+{{- end -}}
```
Comment (igorburago) [inline layouts/partials/bib/entry/extras.html]: Move the whole `with`-block for `.entry.code` before the one for
`.entry.doi`, i.e., move lines 17 through 19 to just above line 14.

I think, it is better to keep the Release link (when it is present)
the last one before the Abstract and BibTeX buttons. This will also
put the Code link next to the Poster and Slide Deck links (if present),
bringing all supplemental material links closer together.

And please fixup these changes into the existing commit and update
its subject line to follow the [recommended commit message style][s].

[s]: https://github.com/uci-iasl/website/pull/32#issuecomment-674346498

**Review by igorburago — CHANGES_REQUESTED** [review]

**Comment by igorburago** [comment]
Have you taken an inspiration for the `code` field name from any
of the existing BibTeX processors or styles, or is it our own custom
name?

It is okay either way, and I am in favor of the feature, but, in the
latter case, I will need to spend a little bit of time researching
whether a similar extension practice already exists, before merging
this variant: It is always better to adopt an existing customization
scheme, unless there is none (as is the case here, I think?) or there
is a clear reason not to (which is not the case here).

**Comment by dcallega** [comment]
The idea here was to introduce a field unique to our project. As far as I know there is no "code" field in standard bibtex files.
I just tried to keep it clean, and I when I could not find any conflict, I used such keyword.

**Comment by dcallega** [comment]
I applied the requested changes and squashed the commits on this feature in one.
I tried to apply the new knowledge, but being the first-ish thought out commit, it might still be suboptimal.
Feedback as always is greatly appreciated!
Once this is closed I will move to #34.

**Comment by igorburago** [comment]
It looks like the class name is still `pubref`. Could you please change it [as suggested](https://github.com/uci-iasl/website/pull/33/files#r470938490)?

**Comment by igorburago** [comment]
I would also recommend changing the commit subject like so:

> Support source code links in *bibliography entries*

I would say that `bibentry:` prefix is unnecessary here.
My logic of using the prefixes in commit messages in this repository
is as follows:

- **No prefix:** Changes to the layout templates and styling of the
website, i.e., anything that Hugo uses to generate the pages, *except
for the content itself* (e.g., Markdown files, PDFs, bibliography data,
images, etc.).
- **Site section-based prefix:** Changes to the content of the site.
The prefix is either a section name alone—`updates`, `people`,
`projects`, `publications` (same as the names of subfolders in
the `content` directory)—or a section name followed by slash and
a subsection—e.g., `people/dcallega` or `projects/city-iot`—when
the commit pertains only the files of that subsection. For changes
in bibliography data, I use a pseudo-section name `bib`.
- **File path-based prefix:** Changes to any site-related external
tools, not used directly to build the website, such as documentation
(beyond the README file), scripts, etc., e.g., `scripts/bibtex2yaml`
or `scripts/gethugo`.

**Comment by igorburago** [comment]
> As far as I know there is no "code" field in standard bibtex files.

That is correct. The reason I asked is that, for some extensions,
many modern BibTeX processors, such as BibLaTeX, already employ
some non-standard fields. For example, to link to digital preprints,
BibLaTeX uses the `eprint`, `eprinttype`, and `eprintclass` fields,
so one should not invent one’s own names for those pieces of
information.

I am not aware of any existing BibTeX extension field being used
for linking to source code, either, so, I think, we will go with `code`.
But I will take another quick look first.

**Comment by dcallega** [comment]
Please let me know if further edits are required.
I squashed the new edits, integrating them into the original commit (edited).

**Comment by dcallega** [comment]
> It looks like the class name is still `pubref`. Could you please change it [as suggested](https://github.com/uci-iasl/website/pull/33/files#r470938490)?

Ops. Done

**Comment by dcallega** [comment]
> I would also recommend changing the commit subject like so:
> 
> > Support source code links in bibliography entries
> 
> I would say that `bibentry:` prefix is unnecessary here.
> My logic of using the prefixes in commit messages in this repository
> is as follows:
> 
> * **No prefix:** Changes to the layout templates and styling of the
>   website, i.e., anything that Hugo uses to generate the pages, _except
>   for the content itself_ (e.g., Markdown files, PDFs, bibliography data,
>   images, etc.).
> * **Site section-based prefix:** Changes to the content of the site.
>   The prefix is either a section name alone—`updates`, `people`,
>   `projects`, `publications` (same as the names of subfolders in
>   the `content` directory)—or a section name followed by slash and
>   a subsection—e.g., `people/dcallega` or `projects/city-iot`—when
>   the commit pertains only the files of that subsection. For changes
>   in bibliography data, I use a pseudo-section name `bib`.
> * **File path-based prefix:** Changes to any site-related external
>   tools, not used directly to build the website, such as documentation
>   (beyond the README file), scripts, etc., e.g., `scripts/bibtex2yaml`
>   or `scripts/gethugo`.

Ok, now I see the pattern! Sorry I probably didn't read enough of them to see it.
Do we want to copy/paste this to the README or even create a small section for "Contribution style"?
I really like this, but I would not be able to explain it myself and we need to do so as we distribute the load to others.

**Comment by igorburago** [comment]
Could you please expand the term `bibentry` in the commit message
to `bibliography entries` or something similar in plain English? E.g.,

> Support source code links in *bibliography entries*

It is not an established term in the repository lingo, and unless we
want to establish it as one, it is better not to use it as such.

I know it is not hard at all to figure out what it means, but it makes
it easier for people to understand and contribute later, when there is
less ad-hoc abbreviations and unidiomatic terms being used in commit
history. I learned that it is a good rule in general from seeing how
things are done in the repositories of bigger collaborative projects.

**Comment by igorburago** [comment]
Okay, I have adjusted the commit message myself.

I also excluded the `code` field from automatically generated BibTeX
snippets, since this customization is not supported by any other
BibTeX processor, and the user copying the snippet will not be able
to benefit from it, anyway.

**Review by igorburago — APPROVED** [review]

**Review by igorburago — APPROVED** [review]
